src/Symfony/Component/Cache/Adapter/FilesystemTagAwareAdapter.php

```diff
@@ -0,0 +1,164 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Cache\Adapter;
+
+use Symfony\Component\Cache\Exception\LogicException;
+use Symfony\Component\Cache\Marshaller\DefaultMarshaller;
+use Symfony\Component\Cache\Marshaller\MarshallerInterface;
+use Symfony\Component\Cache\PruneableInterface;
+use Symfony\Component\Cache\Traits\FilesystemTrait;
+use Symfony\Component\Filesystem\Filesystem;
+
+/**
+ * Stores tag id <> cache id relationship as a symlink, and lookup on invalidation calls.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ * @author André Rømcke <andre.romcke+symfony@gmail.com>
+ *
+ * @experimental in 4.3
+ */
+class FilesystemTagAwareAdapter extends AbstractTagAwareAdapter implements PruneableInterface
+{
+    use FilesystemTrait {
+        doSave as doSaveCache;
+        doDelete as doDeleteCache;
+    }
+
+    /**
+     * Folder used for tag symlinks.
+     */
+    private const TAG_FOLDER = 'tags';
```

Comment:
do we need a separate folder? can't we reuse the existing hashing logic/code/directories instead?

Reply:
makes it much simpler to know that the folder is for when browsing cache directory.

Reply:
ok, then I'd suggest reusing `FilesystemCommonTrait::getFile()` inside/instead `getTagFolder`/`getTagLink`.

Adding a new `string $directory = null` argument to `getFile()` should do it, isn't it?

Reply:
Almost, how should I deal with tagid vs key id?

Reply:
by passing the prefix already concatenated: `$this->getFile(TAGS_PREFIX.$tagId)`?

Reply:
That can work.

But it strictly speaking won't be a file anymore for tag id folder ;) But if you prefer it that way I'm fine.

Reply:
I prefer sharing the hashing code yes

Reply:
I think I found a good way to do what you had in mind 👍